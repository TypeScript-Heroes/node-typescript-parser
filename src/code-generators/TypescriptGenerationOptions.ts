/**
 * Typescript generation options type. Contains all information needed to stringify some objects to typescript.
 *
 * @export
 * @interface TypescriptGenerationOptions
 */
export interface TypescriptGenerationOptions {
    /**
     * Which quote type should be used (' or ").
     *
     * @type {string}
     * @memberof TypescriptGenerationOptions
     */
    stringQuoteStyle: string;

    /**
     * Defines end of line character (semicolon or nothing).
     *
     * @type {('' | ';')}
     * @memberof TypescriptGenerationOptions
     */
    eol: '' | ';';

    /**
     * Defines if the symbols should have spacing in the braces ({ Foo } or {Foo}).
     *
     * @type {boolean}
     * @memberof TypescriptGenerationOptions
     */
    spaceBraces: boolean;

    /**
     * The wrapping method to be used in multiline imports.
     *
     * @type {'MULTIPLE_IMPORTS_PER_LINE' | 'ONE_IMPORT_PER_LINE'}
     * @memberof TypescriptGenerationOptions
     */
    multiLineWrapMethod: 'ONE_IMPORT_PER_LINE' | 'MULTIPLE_IMPORTS_PER_LINE';

    /**
     * The threshold where an import is written as multiline.
     *
     * @type {number}
     * @memberof TypescriptGenerationOptions
     */
    multiLineWrapThreshold: number;

    /**
     * Defines if the last line of a multiline import should have a comma.
     *
     * @type {boolean}
     * @memberof TypescriptGenerationOptions
     */
    multiLineTrailingComma: boolean;

    /**
     * How many spaces of indentiation.
     *
     * @type {number}
     * @memberof TypescriptGenerationOptions
     */
    tabSize: number;
}
