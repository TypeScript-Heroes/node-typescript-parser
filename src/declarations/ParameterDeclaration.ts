import { TypedDeclaration } from './Declaration';

/**
 * Parameter declaration. Is contained in a method or function delaration since a parameter can not be exported
 * by itself.
 *
 * @export
 * @class ParameterDeclaration
 * @implements {TypedDeclaration}
 */
export class ParameterDeclaration implements TypedDeclaration {
    constructor(public name: string, public type: string | undefined, public start?: number, public end?: number) { }
}

export class ObjectBoundParameterDeclaration extends ParameterDeclaration {
    public parameters: ParameterDeclaration[] = [];
    public typeReference: string | undefined;

    public get name(): string {
        return `{ ${this.parameters.map(p => p.name).join(', ')} }`;
    }

    public set name(_: string) { }

    public get type(): string {
        return `{ ${this.parameters.map(p => p.type).join(', ')} }`;
    }

    public set type(_: string) { }

    constructor(start?: number, end?: number) {
        super('', '', start, end);
    }
}

export class ArrayBoundParameterDeclaration extends ParameterDeclaration {
    public parameters: ParameterDeclaration[] = [];
    public typeReference: string | undefined;

    public get name(): string {
        return `[ ${this.parameters.map(p => p.name).join(', ')} ]`;
    }

    public set name(_: string) { }

    public get type(): string {
        return `[ ${this.parameters.map(p => p.type).join(', ')} ]`;
    }

    public set type(_: string) { }

    constructor(start?: number, end?: number) {
        super('', '', start, end);
    }
}
