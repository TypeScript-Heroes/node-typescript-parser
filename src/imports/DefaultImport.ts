import { AliasedImport } from './Import';

/**
 * Default import. Imports the default exports of a file.
 * (i.e. import foobar from ...).
 * 
 * @export
 * @class DefaultImport
 * @implements {AliasedImport}
 */
export class DefaultImport implements AliasedImport {
    public get isNew(): boolean {
        return this.start === undefined || this.end === undefined;
    }

    constructor(public libraryName: string, public alias: string, public start?: number, public end?: number) { }

    /**
     * Clone the current import object.
     * 
     * @returns {DefaultImport}
     * 
     * @memberof DefaultImport
     */
    public clone(): DefaultImport {
        return new DefaultImport(this.libraryName, this.alias, this.start, this.end);
    }
}
