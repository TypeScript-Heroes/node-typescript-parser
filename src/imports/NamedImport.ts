import { SymbolSpecifier } from '../SymbolSpecifier';
import { Import } from './Import';

/**
 * Basic typescript import (ES6 style). Does contain multiple symbols of a file and converts
 * itself to a multiline import if the threshold is reached.
 * (i.e. import {Foobar} from ...).
 * 
 * @export
 * @class NamedImport
 * @implements {Import}
 */
export class NamedImport implements Import {
    public specifiers: SymbolSpecifier[] = [];

    public get isNew(): boolean {
        return this.start === undefined || this.end === undefined;
    }

    constructor(public libraryName: string, public start?: number, public end?: number) { }

    /**
     * Clone the current import object.
     * 
     * @returns {NamedImport}
     * 
     * @memberof NamedImport
     */
    public clone(): NamedImport {
        const clone = new NamedImport(this.libraryName, this.start, this.end);
        clone.specifiers = this.specifiers.map(o => o.clone());
        return clone;
    }
}
