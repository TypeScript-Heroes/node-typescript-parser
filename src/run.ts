import { TypescriptParser } from './';

const parser = new TypescriptParser();

const code = `
class Foo {
    private _calc: string;

    public get calc(): string {
        return this._calc;
    }

    public set calc(value: string) {
        this._calc = value;
    }

    public foo(): void {

    }
}
`;

async function parse(): Promise<void> {
    // const parsed = await parser.parseSource(`import Foo, { Foobar } from 'myLib';`);
    // const parsed = await parser.parseSource(`import Foo from 'myLib';`);
    // const parsed = await parser.parseSource(`import { Foo } from 'myLib';`);
    // const parsed = await parser.parseSource(`import { Foo, default as Bar } from 'myLib';`);
    const parsed = await parser.parseSource(code);
    console.log(parsed);
}

parse().then(() => process.exit(0)).catch(() => process.exit(1));
