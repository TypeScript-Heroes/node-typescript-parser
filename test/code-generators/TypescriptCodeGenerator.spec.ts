import { TypescriptCodeGenerator } from '../../src/code-generators/TypescriptCodeGenerator';
import { TypescriptGenerationOptions } from '../../src/code-generators/TypescriptGenerationOptions';
import { ClassDeclaration } from '../../src/declarations';
import { DeclarationVisibility } from '../../src/declarations/DeclarationVisibility';
import { MethodDeclaration } from '../../src/declarations/MethodDeclaration';
import { ParameterDeclaration } from '../../src/declarations/ParameterDeclaration';
import { PropertyDeclaration } from '../../src/declarations/PropertyDeclaration';
import { VariableDeclaration } from '../../src/declarations/VariableDeclaration';
import { NotGeneratableYetError } from '../../src/errors/NotGeneratableYetError';
import { ExternalModuleImport } from '../../src/imports/ExternalModuleImport';
import { NamedImport } from '../../src/imports/NamedImport';
import { NamespaceImport } from '../../src/imports/NamespaceImport';
import { StringImport } from '../../src/imports/StringImport';
import { SymbolSpecifier } from '../../src/SymbolSpecifier';

const namedImport = new NamedImport('namedLib');
const multiLineNamedImport = new NamedImport('multiLineNamedLib');

namedImport.specifiers = [
    new SymbolSpecifier('spec1'),
    new SymbolSpecifier('spec2', 'alias2'),
];

multiLineNamedImport.specifiers = [
    new SymbolSpecifier('spec1'),
    new SymbolSpecifier('spec2'),
    new SymbolSpecifier('spec3'),
    new SymbolSpecifier('spec4'),
    new SymbolSpecifier('spec5'),
    new SymbolSpecifier('spec6'),
    new SymbolSpecifier('spec7'),
    new SymbolSpecifier('spec8'),
    new SymbolSpecifier('spec9'),
    new SymbolSpecifier('spec10'),
    new SymbolSpecifier('spec11'),
    new SymbolSpecifier('spec12'),
    new SymbolSpecifier('spec13'),
    new SymbolSpecifier('spec14'),
    new SymbolSpecifier('spec15'),
];


describe('TypescriptCodeGenerator', () => {
    const defaultOptions: TypescriptGenerationOptions = {
        eol: ';',
        multiLineTrailingComma: true,
        multiLineWrapThreshold: 125,
        spaceBraces: true,
        stringQuoteStyle: `'`,
        tabSize: 4,
    };
    const generatables = [
        new SymbolSpecifier('SymbolSpecifier'),
        new SymbolSpecifier('SymbolSpecifier', 'WithAlias'),
        new MethodDeclaration('myMethod', false, DeclarationVisibility.Public, 'void'),
        new MethodDeclaration('myAbstractMethod', true, DeclarationVisibility.Public, 'string'),
        new MethodDeclaration('myProtectedMethod', false, DeclarationVisibility.Protected, 'void'),
        new MethodDeclaration('myProtectedAbstractMethod', true, DeclarationVisibility.Protected, 'string'),
        new MethodDeclaration('myPrivateMethod', false, DeclarationVisibility.Private, 'void'),
        new MethodDeclaration('myPrivateAbstractMethod', true, DeclarationVisibility.Private, 'string'),
        new ParameterDeclaration('param', ''),
        new ParameterDeclaration('stringParam', 'string'),
        new PropertyDeclaration('pubProperty', DeclarationVisibility.Public, 'string'),
        new PropertyDeclaration('protProperty', DeclarationVisibility.Protected, 'number'),
        new PropertyDeclaration('prvProperty', DeclarationVisibility.Private, 'boolean'),
        new VariableDeclaration('myVar', false, false, 'string'),
        new VariableDeclaration('myConst', true, false, 'string'),
        new ExternalModuleImport('externalModuleLib', 'externalAlias'),
        new StringImport('stringLib'),
        new NamespaceImport('namespaceLib', 'namespaceAlias'),
        namedImport,
        multiLineNamedImport,
        new NamedImport('emptyImport'),
    ];

    for (const generatable of generatables) {

        it(`should generate the correct code for ${generatable.constructor.name}`, () => {
            const generator = new TypescriptCodeGenerator(defaultOptions);

            expect(generator.generate(generatable)).toMatchSnapshot();
        });

    }

    it('should throw on non generatable element', () => {
        const generator = new TypescriptCodeGenerator(defaultOptions);

        expect(() => generator.generate(new ClassDeclaration('foo', true))).toThrow(NotGeneratableYetError);
    });
});
