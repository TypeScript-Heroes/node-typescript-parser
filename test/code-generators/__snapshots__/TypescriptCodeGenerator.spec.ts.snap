// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypescriptCodeGenerator should generate the correct code for ExternalModuleImport 1`] = `"import externalAlias = require('externalModuleLib');"`;

exports[`TypescriptCodeGenerator should generate the correct code for MethodDeclaration 1`] = `
"    public myMethod(): void {
        throw new Error('Not implemented yet.');
    }
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for MethodDeclaration 2`] = `
"    public myAbstractMethod(): string {
        throw new Error('Not implemented yet.');
    }
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for MethodDeclaration 3`] = `
"    protected myProtectedMethod(): void {
        throw new Error('Not implemented yet.');
    }
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for MethodDeclaration 4`] = `
"    protected myProtectedAbstractMethod(): string {
        throw new Error('Not implemented yet.');
    }
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for MethodDeclaration 5`] = `
"    private myPrivateMethod(): void {
        throw new Error('Not implemented yet.');
    }
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for MethodDeclaration 6`] = `
"    private myPrivateAbstractMethod(): string {
        throw new Error('Not implemented yet.');
    }
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for NamedImport 1`] = `"import { spec1, spec2 as alias2 } from 'namedLib';"`;

exports[`TypescriptCodeGenerator should generate the correct code for NamedImport 2`] = `
"import {
    spec1,
    spec10,
    spec11,
    spec12,
    spec13,
    spec14,
    spec15,
    spec2,
    spec3,
    spec4,
    spec5,
    spec6,
    spec7,
    spec8,
    spec9,
} from 'multiLineNamedLib';"
`;

exports[`TypescriptCodeGenerator should generate the correct code for NamedImport 3`] = `"import { } from 'emptyImport';"`;

exports[`TypescriptCodeGenerator should generate the correct code for NamedImport 4`] = `"import Default from 'defaultImport';"`;

exports[`TypescriptCodeGenerator should generate the correct code for NamedImport 5`] = `"import Default, { spec1, spec2 as alias2 } from 'defaultWithNamedImport';"`;

exports[`TypescriptCodeGenerator should generate the correct code for NamedImport 6`] = `
"import Default, {
    spec1,
    spec10,
    spec11,
    spec12,
    spec13,
    spec14,
    spec15,
    spec2,
    spec3,
    spec4,
    spec5,
    spec6,
    spec7,
    spec8,
    spec9,
} from 'defaultWithNamedMultilineImport';"
`;

exports[`TypescriptCodeGenerator should generate the correct code for NamespaceImport 1`] = `"import * as namespaceAlias from 'namespaceLib';"`;

exports[`TypescriptCodeGenerator should generate the correct code for ParameterDeclaration 1`] = `"param"`;

exports[`TypescriptCodeGenerator should generate the correct code for ParameterDeclaration 2`] = `"stringParam: string"`;

exports[`TypescriptCodeGenerator should generate the correct code for PropertyDeclaration 1`] = `
"    public pubProperty: string;
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for PropertyDeclaration 2`] = `
"    protected protProperty: number;
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for PropertyDeclaration 3`] = `
"    private prvProperty: boolean;
"
`;

exports[`TypescriptCodeGenerator should generate the correct code for StringImport 1`] = `"import 'stringLib';"`;

exports[`TypescriptCodeGenerator should generate the correct code for SymbolSpecifier 1`] = `"SymbolSpecifier"`;

exports[`TypescriptCodeGenerator should generate the correct code for SymbolSpecifier 2`] = `"SymbolSpecifier as WithAlias"`;

exports[`TypescriptCodeGenerator should generate the correct code for VariableDeclaration 1`] = `"myVar: string"`;

exports[`TypescriptCodeGenerator should generate the correct code for VariableDeclaration 2`] = `"myConst: string"`;
